pkg/controller/core.oam.dev/v1alpha2/application/revision.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2021 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package application
+
+import (
+	"context"
+
+	"github.com/google/go-cmp/cmp"
+
+	"github.com/pkg/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha2"
+	"github.com/oam-dev/kubevela/pkg/appfile"
+	"github.com/oam-dev/kubevela/pkg/controller/utils"
+)
+
+// GenerateRevision will generate revision for an Application when created/updated
+func (r *Reconciler) GenerateRevision(ctx context.Context, appfile *appfile.Appfile, application *v1alpha2.Application, ac *v1alpha2.ApplicationConfiguration, comps []*v1alpha2.Component) (bool, *v1alpha2.AppRevisionHash, error) {
```

Comment:
The names of comps being passed in are not versioned. Should they be?

Reply:
Also note that the ac passed in here will use the ComponentName instead of the revisionName.
The revisionName is only modified after apply():

```go
func (h *appHandler) apply(ctx context.Context, ac *v1alpha2.ApplicationConfiguration, comps []*v1alpha2.Component) error {

	ac.SetOwnerReferences(owners)
	for _, comp := range comps {
		...
		revisionName, err := h.createOrUpdateComponent(ctx, newComp)
		if err != nil {
			return err
		}
		// find the ACC that contains this component
		for i := 0; i < len(ac.Spec.Components); i++ {
			// update the AC using the component revision instead of component name
			// we have to make AC immutable including the component it's pointing to
			if ac.Spec.Components[i].ComponentName == newComp.Name {
				ac.Spec.Components[i].RevisionName = revisionName
				ac.Spec.Components[i].ComponentName = ""
			}
		}
	}
```

Reply:
yes, this will cause problems. We compute the revision with AC pointing to component (mutable) when the real AC created pointing to componentRevision (imutable)